### ISSUE TITLE
Relevance-Promoting Language Model for Short-Text Conversation

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#8]
# 논문 정보 (URL, 저자 등)

[https://arxiv.org/pdf/1911.11489.pdf](https://arxiv.org/pdf/1911.11489.pdf)

The Chinese University of Hong Kong

Tencent AI Lab

# 개요

- Short Text Conversation (STC) 태스크에서는 Seq2Seq 프레임워크가 널리 쓰이고 있습니다.
    - 하지만, 이 프레임워크에는 본질적인 문제들이 존재합니다.
        1. auto-regressive decoder가 ground truth response text의 로스로 학습되기 때문에, 인코더의 입력 query의 영향이 미미하거나, 무시되는 경향이 생깁니다. (컨텍스트에 둔감해짐)
        2. 디코딩시에 주로 사용되는 maximization-based 디코딩 전략(e.g. greedy or beam search)은 자주 등장하는 phrases 위주로 검색 공간이 제한되기 때문에 likelihood가 비정상적으로 높은 답변들이 생성되곤 합니다.
- 이러던 도중, GPT-2가 등장했습니다.
    - GPT-2의 이점은,  임의의 길이를 갖는 context를 입력받을 수 있을 뿐만 아니라, 모델이 디코더 그 자체이기 때문에 textual supervision 신호를 최대한으로 활용할 수 있습니다.
    - 또한, top-k 샘플링을 이용하기 때문에, 다양한 문장을 생성할 수 있습니다.
- 이러한, Language 모델은 이점도 있지만, 단점도 있습니다.
    - STC 태스크와 같은 conditional한 생성 태스크에서는 query 문장과 target 문장이 concat되어 입력되기 때문에 두 문장을 구분하기 어려워집니다.
    - 또한, STC 태스크에 언어 모델을 도입할 때의 또 다른 포텐셜 이슈가 있는데,
        1. **recency bias (Khandelwal et al. 2018)**
        2. **explanation-away bias (Yu et al. 2017)**
        - 이 문제들은 언어 모델이 거의 직전 컨텍스트에만 의존하게 되고, long-term context를 경시하게 된다는 이슈입니다. (언어 모델에서의 long-term context는 seq2seq에서의 source information과 거의 동일해집니다.)
        - 따라서, fluent한 문장은 생성되지만, 토픽적으로 관계없는 문장이 생성되곤 합니다.
- 본 논문에서는 위와 같은 문제들을 해결하기 위해서, 다음과 같은 전략을 제시합니다.
    - **Supervised Source Attention (SSA)** 컴포넌트와 **Topic Inference (TI)** 컴포넌트를 이용하여 언어 모델에서의 relevance 모델링을 강화합니다.
    - **randomization-over-maximization** 디코딩 전략을 이용하여 다양한 문장을 생성합니다.
        - 이 전략은 top-k sampling이랑 완전히 동일하나, 시기상 top-k 샘플링이 유행하기 전에 나온 논문인듯합니다.

# 모델

<img width="480" alt="스크린샷 2020-08-31 오후 6 41 35" src="https://user-images.githubusercontent.com/49383323/91706696-afda0700-ebb9-11ea-8a28-371cd582a6c3.png">

- 이 논문에서 풀고자 하는 objective는, **X**={**x**1,...,**x**m, **x**m+1,...,**x**n}이라는 싱글 시퀀스에서 **x**1:m이 쿼리, **x**m+1:n가 답변일 때, P(**x**1:n)라는 unconditional likelihood를 최대화하는 것입니다.
    - Meena에서 P(**x**m+1:n)를 최대화하는 것과는 다른 objective입니다.
    - 직관적으로 response 문장만을 생성하는 Seq2Seq 프레임워크보다 query 문장의 생성도 학습 과정에 포함시키는 언어 모델 프레임워크가 더 fluent한 문장을 생성할 수 있을거라 생각할 수 있습니다.
- 모델 구조는 decoder-only 트랜스포머 구조이고, 추가적으로 SSA, TI 컴포넌트가 붙어있습니다.

<img width="706" alt="스크린샷 2020-08-31 오후 6 41 45" src="https://user-images.githubusercontent.com/49383323/91706690-ad77ad00-ebb9-11ea-8687-6baed6771a73.png">

### Supervised Source Attention

- 기존의 Seq2Seq 프레임워크에서는 query/source information이 encoder와 decoder 사이의 attention (q를 decoder representations, k,v를 encoder representations로 하는 어텐션)으로 결합이 되어있습니다.
    - 이것과 유사하게, 언어 모델의 입력 시퀀스 중 컨텍스트에 대해서만 어텐션을 주는 것이 relevance의 관점에서는 도움이 될 것입니다.
    - 따라서, 각 레이어의 self-attention layer의 상단에 source attention layer를 추가하는 전략을 이용합니다.
    - 단, 이 어텐션은 현재 step이 reply token일 경우에만 적용했습니다.
- word alignment를 학습할 때는, supervision이나 외부 지식을 이용하는 것이 더 도움이 될 것입니다.
    - 우리는 쿼리의 특정 토큰들을 외부 지식으로 간주하고, source attention이 특정 쿼리 토큰에 집중할 수 있도록 합니다.
        - beta는 i번째 소스 토큰에 대한 m번째 쿼리 토큰의 attention score

   <img width="412" alt="스크린샷 2020-08-31 오후 6 43 57" src="https://user-images.githubusercontent.com/49383323/91706857-ef085800-ebb9-11ea-9683-72a8e4d8cb24.png">

    - 아래와 같이 query keyword indicator vector (**y**_src)와의 Euclidean Distance 로스를 추가하여 학습합니다.
        - 단, 특정 토큰이 keyword인지 아닌지는 외부 툴킷을 이용해서 미리 레이블링해둔 값입니다.

<img width="352" alt="스크린샷 2020-08-31 오후 6 43 45" src="https://user-images.githubusercontent.com/49383323/91706901-00e9fb00-ebba-11ea-90de-b67d1c847c86.png">

### Topic Inference

- SSA는 특정 중요한 쿼리 토큰에 강조를 둬서, relevance를 향상시키겠다는 전략입니다.
- 하지만, 쿼리와 토픽적으로 연관이 있을만한 단어의 범위는 토픽에 명시적으로 주어진 키워드보다 훨씬 범위가 넓습니다.
    - 예컨대, "제일 좋아하는 **과일**이 뭐야?"라는 쿼리에 대해서 나올 수 있는 답변들은 "**수박**을 제일 좋아해", "내가 제일 좋아하는 건 **파인애플**이야" 등이 있을 수 있습니다.
- 따라서, 본 논문에서는 훈련셋의 특정 쿼리에 대한 여러개의 정답 답변들로부터 키워드를 추출했습니다.
    - 이 키워드들이 아래에서 global keyword indicator로 사용됩니다.
- latent topic information을 추론하기 위해서, 아래와 같은 식으로 query representation에 대한 global topical word distribution을 추정합니다.

   <img width="445" alt="스크린샷 2020-08-31 오후 6 44 45" src="https://user-images.githubusercontent.com/49383323/91706925-0d6e5380-ebba-11ea-962a-f2741d612738.png">

    - 위에서, last query hidden representation을 함수 f에 태운 값이 **h**q가 됩니다.
        - 함수 **f**: tanh-activated linear layer
        - z는 vocab 개수만큼 존재
- 아래와 같이 global keyword indicator vector **y**kwd와의 cross-entropy를 추가적인 로스로 사용했습니다.
    - 단, 일반적으로 multi-label classification에서 자주 쓰이는 component-wise한 sigmoid로도 (6)식을 변경해보았으나, 성능이 더 떨어져 softmax를 이용했다고 합니다.

   <img width="404" alt="스크린샷 2020-08-31 오후 6 45 27" src="https://user-images.githubusercontent.com/49383323/91707026-31319980-ebba-11ea-81e6-7418f3cfbecc.png">

- 또한, 위의 **h**q값은 아래와 같이 gate 함수에 사용합니다.

<img width="452" alt="스크린샷 2020-08-31 오후 6 45 32" src="https://user-images.githubusercontent.com/49383323/91707060-3d1d5b80-ebba-11ea-9cb3-3439c60d1906.png">

### 학습

- 결과적으로 학습은 다음과 같은 식을 최적화하는 방식으로 진행됩니다.

<img width="428" alt="스크린샷 2020-08-31 오후 6 45 38" src="https://user-images.githubusercontent.com/49383323/91708071-bff2e600-ebbb-11ea-965c-ccedd42251e0.png">

### Decoding

- 디코딩 시에는 top-k sampling 방식을 이용합니다.

# 데이터

- benchmark STC 중국어 데이터셋을 이용합니다.
    - Weibo의 실제 대화를 베이스로 한 7M 사이즈의 query-response pairs입니다.
- 토크나이징은 문자 단위 (한자), 12000 사이즈의 보캡을 이용합니다.

# 실험

### Evaluation Metrics

- Relevance Metrics
    - BLEU-2,3,4
    - HIT-Q, HIT-R: 쿼리와 답변의 topical 단어의 hit rates
        - topical 단어는 외부 툴킷을 이용해서 미리 추출해둔 키워드 셋을 사용합니다.
- Diversity Metrics
    - DIST-1,2 : 생성 답변의 유니크한 uni-grams, bi-grams 비율
- 정성 평가
    - Relevance는 0-3, Fluency는 0-2, Acceptance는 0-1로 평가했다고 합니다.

### 베이스라인

- LSTM-LM: LSTM Language model
- LSTM-S2S: LSTM Encoder-Decoder model
- TFM-S2S: Transformer Encoder-Decoder
- TFM-LM: Transformer Language Model
- MMI: LSTM-S2S를 MMI 리랭킹
- CVAE: Conditional Variational Auto-Encoder
- MMPMS (Chen et al. 2019): STC 태스크의 SOTA 모델

### 결과

<img width="1287" alt="_2020-04-14__4 25 36" src="https://user-images.githubusercontent.com/49383323/91708162-e31d9580-ebbb-11ea-9c75-8550918d36f4.png">

<img width="761" alt="_2020-04-14__4 27 47" src="https://user-images.githubusercontent.com/49383323/91708188-ec0e6700-ebbb-11ea-905d-380ec02f680d.png">

- 주목할만한건, TFM-S2S가 Relevance와 Acceptance관점에서는 더 높은 성능을 보여주고 있으나, Fluency는 LM 모델이 더 좋은 성능을 보여주고 있습니다.
    - 즉, 앞에서 언급했듯이, LM 모델은 컨텍스트의 생성도 학습 과정에 포함되기 때문에 전체적으로 fluent한 문장 생성은 더 잘하지만, long-term context를 고려하지 못한 답변 생성이 되기 쉽다는 것을 의미합니다.
- 또한 SSA 컴포넌트를 빼는게 오히려 Fluency 관점에서는 더 성능이 좋았습니다.
    - 이것은, query context이 언어 모델링 (다음 단어 생성)하는 관점에서는 도움이 되지 않는다는 것을 암시합니다.

### Top-k Sampling의 효과

- 실제 생성 문장을 확인해보면, Beam Search를 이용할 경우, 특정 토큰 반복 문제나 쿼리를 그대로 copy해오는 문제가 있었습니다.
    - 여기서는, 쿼리와 답변사이의 겹치는 문자열의 길이가 4이상이면 copy로 간주하고, 동일 word/phrase의 루프가 3번이상 반복되면 repetition으로 간주했습니다.

<img width="752" alt="_2020-04-14__4 42 14" src="https://user-images.githubusercontent.com/49383323/91708203-f4ff3880-ebbb-11ea-90f6-17b8aaad0924.png">

# 논의할만한 내용

- 논문에서 제시하는 방법론이 복잡성에 비해 갖는 성능 개선이 미미함.
- 다만, LM 기반 생성 모델, Seq2Seq 기반 생성 모델의 차이에 대한 인사이트는 생김.

# 관련된 논문
